### PR TITLE
Implement loss recovery from the current WG draft

### LINKED ISSUE #498: Reimplement loss recovery from the WG draft

### PR DESCRIPTION
**Pull request by lucas-clemente**
Fixes #498 and will hopefully go a long way towards fixing the many
flaky tests.

### CODE REVIEW

**Comment by marten-seemann** [comment]
I think we need to check if a packet is retransmittable in `sentPacketHandler.SentPacket()` (it probably makes sense to define this as a method on the `Packet`.

**Comment by lucas-clemente** [comment]
Agreed, but I'd rather do this as a separate PR later. Filed #505.

File: ackhandler/sent_packet_handler.go [hunk 1]
```diff
@@ -12,6 +12,18 @@ import (
 	"github.com/lucas-clemente/quic-go/utils"
 )
 
+const (
+	// Maximum reordering in time sapce before time based loss detection considers a packet lost.
```
Comment (marten-seemann) [inline ackhandler/sent_packet_handler.go]: Nit: typo.

File: ackhandler/sent_packet_handler.go [hunk 2]
```diff
@@ -12,6 +12,18 @@ import (
 	"github.com/lucas-clemente/quic-go/utils"
 )
 
+const (
```
Comment (marten-seemann) [inline ackhandler/sent_packet_handler.go]: Maybe move all the Loss Recovery constants to a separate file.

File: ackhandler/sent_packet_handler.go [hunk 3]
```diff
@@ -211,59 +204,124 @@ func (h *sentPacketHandler) ReceivedAck(ackFrame *frames.AckFrame, withPacketNum
...
+		// Packets are sorted by number, so we can stop searching
+		if packet.PacketNumber > largestAcked {
+			break
+		}
 	}
+	return false
+}
 
-	h.garbageCollectSkippedPackets()
+func (h *sentPacketHandler) updateLossDetectionAlarm() {
```
Comment (marten-seemann) [inline ackhandler/sent_packet_handler.go]: Why did you rename `setLossDetectionAlarm()` from the WG draft?

File: ackhandler/sent_packet_handler.go [hunk 4]
```diff
@@ -211,59 +204,124 @@ func (h *sentPacketHandler) ReceivedAck(ackFrame *frames.AckFrame, withPacketNum
...
+	// Cancel the alarm if no packets are outstanding
+	if h.packetHistory.Front() == nil {
```
Comment (marten-seemann) [inline ackhandler/sent_packet_handler.go]: Better: 
```go
if h.packetHistory.Len() == 0
```

File: ackhandler/sent_packet_handler.go [hunk 5]
```diff
@@ -211,59 +204,124 @@ func (h *sentPacketHandler) ReceivedAck(ackFrame *frames.AckFrame, withPacketNum
...
+	}
+}
 
-	h.congestion.OnCongestionEvent(
-		rttUpdated,
-		h.BytesInFlight(),
-		ackedPackets,
-		lostPackets,
-	)
+// TODO(lclemente): Introducing congestion.MaybeExitSlowStart() would allow us to call through for each packet and eliminate both the rttUpdated param and the packet slices passed to the congestion
```
Comment (marten-seemann) [inline ackhandler/sent_packet_handler.go]: Please use your Github username for TODOs.

File: ackhandler/sent_packet_handler.go [hunk 6]
```diff
@@ -211,59 +204,124 @@ func (h *sentPacketHandler) ReceivedAck(ackFrame *frames.AckFrame, withPacketNum
...
+
+		if packet.PacketNumber > h.LargestAcked {
+			break
+		}
+
+		timeSinceSent := now.Sub(packet.SendTime)
+		if timeSinceSent > delayUntilLost {
+			lostPackets = append(lostPackets, el)
+		} else if h.lossTime.IsZero() {
+			// Note: This conditional is only entered once per call
```
Comment (marten-seemann) [inline ackhandler/sent_packet_handler.go]: This note is not obvious to me. Can you add a reason?

File: ackhandler/sent_packet_handler.go [hunk 7]
```diff
@@ -149,54 +138,58 @@ func (h *sentPacketHandler) ReceivedAck(ackFrame *frames.AckFrame, withPacketNum
...
+		return ErrAckForSkippedPacket
+	}
 
-	// check if it acks any packets that were skipped
-	for _, p := range h.skippedPackets {
-		if ackFrame.AcksPacket(p) {
-			return ErrAckForSkippedPacket
+	rttUpdated := h.maybeUpdateRTT(ackFrame.LargestAcked, ackFrame.DelayTime, rcvTime)
+
+	ackedPackets, err := h.determineNewlyAckedPackets(ackFrame)
```
Comment (marten-seemann) [inline ackhandler/sent_packet_handler.go]: I think there's a better way than creating a slice just to loop over it later.

**Comment by marten-seemann** [comment]
Does this resolve the flakiness of our integration tests? If so, I'd be very happy if we could remove the `travis_retry` hack.

File: ackhandler/sent_packet_handler.go (hunk 1 above)
Comment (lucas-clemente) [inline ackhandler/sent_packet_handler.go]: Done

File: ackhandler/sent_packet_handler.go (hunk 2 above)
Comment (lucas-clemente) [inline ackhandler/sent_packet_handler.go]: I don't see a benefit, and they are not too many here. For constants that are only used in one place, I'd rather have them as local ad as close to their use as possible.

File: ackhandler/sent_packet_handler.go (hunk 3 above)
Comment (lucas-clemente) [inline ackhandler/sent_packet_handler.go]: Because we don't set the alarm here, only update one particular alarm (h.alarm) that is then set in the session.

File: ackhandler/sent_packet_handler.go (hunk 4 above)
Comment (lucas-clemente) [inline ackhandler/sent_packet_handler.go]: Done

File: ackhandler/sent_packet_handler.go (hunk 5 above)
Comment (lucas-clemente) [inline ackhandler/sent_packet_handler.go]: 🙄

File: ackhandler/sent_packet_handler.go (hunk 6 above)
Comment (marten-seemann) [inline ackhandler/sent_packet_handler.go]: Actually, this is pretty obvious ;)

File: ackhandler/sent_packet_handler.go (hunk 7 above)
Comment (lucas-clemente) [inline ackhandler/sent_packet_handler.go]: True, this is slightly suboptimal. But I really like this structure, as it makes determineNewlyAckedPackets pure and therefore much easier to test (and isolates concerns between the functions better). I also want to try applying the same pattern to detectLostPackets, but it's a bit trickier to make that pure as well.

File: ackhandler/sent_packet_handler.go (hunk 6 above)
Comment (lucas-clemente) [inline ackhandler/sent_packet_handler.go]: Leaving it as is after OOB discussion.
